### ISSUE TITLE
[Improvement idea] Implement read model optimized for frame decoders

### ISSUE THREAD

**Issue by ninja-** [netty/netty#4314]
In my opinion the read model isn't really optimized for frame decoders.
There should be a way to read directly a frame into a new buffer without both slice() and copy() etc. more like in C

So a frame decoder should read a header directly from the socket, then decide the frame's length, and then read exactly X bytes into new buffer (or keep trying).
 
The current model has following problems:
- forced cumulation if data comes over X reads, copying bytes over and over
- slice() creates per-read gc pressure because the slices aren't pooled so I am not going to use it, no way
- even if I were to use slice(), I couldn't grow the buffer and would have to copy the data to a new one _anyway_
- if more than one slice is used on the stream the other stream data could be kept forever etc. and not discarded
- copy() would copy the bytes without need as the improved model could just read to a new separately buffer directly

**Comment by Scottmitch** [comment]
@ninja- - Are you familiar with [RecvByteBufAllocator](https://github.com/netty/netty/blob/master/transport/src/main/java/io/netty/channel/RecvByteBufAllocator.java) (also see the interfaces / types that inherit from this class)? You can use this to control how to allocate `ByteBuf`s while preforming reads from the underlying transport mechanism. In 4.1+ you can use this to control the read loop too in terms of how many max total bytes should be attempted to read, or how many max read operations should attempted per channel when there is data to read (the later can also be done in older versions of Netty). If this interface is not adequate can you explain why?

**Comment by ninja-** [comment]
@Scottmitch thanks for a tip. do you think it can be useful to create a more low-level decoder? Is it worth the time at all?

for example:
I consider the header to be max 10 bytes including all the pipeline things decryption etc.
would it make sense if I read these 10 bytes and denied reading more if there isn't at least X bytes in the buffer to avoid read cumulations?

But the read cumulations are one thing and the other thing is that the frame decoder needs to copy the data to the new buffer which will be decoded frame. On reflection, I think that this couldn't be optimized given that the pipeline contains other elements before a frame can be decoded (encryption etc.)

**Comment by ninja-** [comment]
I switched to level triggered, it works fine but with one regression: it will starting missing last packets when the remote end disconnects, even if autoread=true. Problem goes away when epoll mode is back to edge triggered. is this a bug?

**Comment by normanmaurer** [comment]
That woukd be a bug... Let me investigate

**Comment by ninja-** [comment]
@normanmaurer any news on this?

**Comment by normanmaurer** [comment]
@ninja- nope not yet... had no time today. Would you mind open a bug-report for this and mention me in it ?

**Comment by ninja-** [comment]
#4313

**Comment by Scottmitch** [comment]
`denied reading more if there isn't at least X bytes in the buffer to avoid read cumulations?`

How will you `deny` reading more data? You are using encryption so how do you know how many total clear text bytes there are after decryption (without reading and doing the decryption)?

**Comment by ninja-** [comment]
@Scottmitch I could get the first 10 bytes, get the 10 bytes through the pipeline to get the header decoded and then decide. But maybe it's too crazy to optimize anything with a visible CPU difference.

**Comment by Scottmitch** [comment]
@ninja- - I'm still not sure I understand. Are your header and body encrypted? How do you correlate number of encrypted bytes (what you read off the wire) to decrypted bytes (what the application sees)?

I guess general advise is to only optimize if there is a need, and analyze your bottlenecks with tools to know where to focus your efforts.

**Comment by ninja-** [comment]
@Scottmitch  yeah I dont think that was a good idea but I am hoping for some improvements after fixing epoll
